jsonschema/_validators.py

```diff
@@ -190,9 +190,14 @@ def enum(validator, enums, instance, schema):
 
 
 def ref(validator, ref, instance, schema):
-    with validator.resolver.resolving(ref) as resolved:
+    scope, resolved = validator.resolver.resolve(ref)
+    validator.resolver.push_scope(scope)
```

Comment:
There's another backwards incompatibility here, which is that these are newly introduced methods on the `RefResolver` interface that people who might have implemented them did not need until now.

To resolve that, I think we'd either need to fall back to `resolving` if there was no `push_scope` attribute, or perhaps you might consider an interface that composes `RefResolver` rather than modifying `RefResolver` itself. Lemme know if this makes sense.